### PR TITLE
add infer type to custom op

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc [hunk 1]
```diff
@@ -2202,14 +2202,23 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxCustomOpRegister
...
-          ret->forward =
-            static_cast<bool(*)(int, void**, int*, const int*, const bool, void*)>(forwardEntry);
-          ret->backward =
-            static_cast<bool(*)(int, void**, int*, const int*, const bool, void*)>(backwardEntry);
-          ret->del = static_cast<bool(*)(void*)>(delEntry);
-          ret->p_forward = state;
-          ret->p_backward = state;
-          ret->p_del = state;
+          // TODO(eric): Memory leak here. Refactor later and delete in delEntry
+          ret->num_callbacks = 3;
```
Comment (piiswrong) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: @javelinjs @Ldpe2G

File: src/executor/graph_executor.cc [hunk 2]
```diff
@@ -733,17 +733,18 @@ void GraphExecutor::InitOpSegs() {
           op_node.exec->exec_type() != Operator::kSync) {
         cached_seg_opr_[topo_start] = this->CreateCachedSegOpr(topo_start, nid);
         topo_start = nid + 1;
-      }
-      // If it produces output gradient, don't include it in the segment
-      bool output_gradient = false;
-      for (auto &out_arr : op_node.exec->out_array) {
-        if (grad_vars.find(out_arr.var()) != grad_vars.end()) {
-          output_gradient = true;
+      } else {
```
Comment (piiswrong) [inline src/executor/graph_executor.cc]: @eric-haibin-lin

Comment (eric-haibin-lin) [inline src/executor/graph_executor.cc]: I see. Good catch! I left out a `continue;` after increment topo_start. This works

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc (hunk 1 above)
Comment (Ldpe2G) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: @piiswrong, will follow after this merge

**Pull request merged**